File: impl/lua/src/main/lua/liff/dictionary.lua
local model = require("liff.model")
local normalizer = require("liff.normalize")

local M = {}

M.PERFECT_SCORE = 1000
M.QUALIFYING_SCORE = 700
M.LOW_SUGGESTION_COUNT = 2
M.TOKEN_PREFIX_SCORE = 900
M.PARTIAL_PREFIX_SCORE = 750
M.PREFIX_MIN_CODE_POINTS = 4

local function code_points(value)
    local result = {}
    for _, code_point in utf8.codes(value) do
        result[#result + 1] = code_point
    end
    return result
end

function M.compare_code_points(left, right)
    local left_points = code_points(left)
    local right_points = code_points(right)
    local common = math.min(#left_points, #right_points)
    for index = 1, common do
        if left_points[index] ~= right_points[index] then
            return left_points[index] < right_points[index] and -1 or 1
        end
    end
    if #left_points == #right_points then
        return 0
    end
    return #left_points < #right_points and -1 or 1
end

function M.damerau_levenshtein(left, right)
    local left_points = code_points(left)
    local right_points = code_points(right)
    local previous_previous = {}
    local previous = {}
    for index = 1, #right_points + 1 do
        previous_previous[index] = 0
        previous[index] = index - 1
    end

    for left_index = 1, #left_points do
        local current = { left_index }
        for right_index = 1, #right_points do
            local column = right_index + 1
            local substitution = left_points[left_index] == right_points[right_index] and 0 or 1
            current[column] = math.min(
                current[column - 1] + 1,
                previous[column] + 1,
                previous[column - 1] + substitution
            )
            if left_index > 1 and right_index > 1
                and left_points[left_index] == right_points[right_index - 1]
                and left_points[left_index - 1] == right_points[right_index]
            then
                current[column] = math.min(
                    current[column],
                    previous_previous[right_index - 1] + 1
                )
            end
        end
        previous_previous = previous
        previous = current
    end
    return previous[#right_points + 1]
end

function M.similarity_score(left, right)
    local maximum = math.max(utf8.len(left), utf8.len(right))
    if maximum == 0 then
        return M.PERFECT_SCORE
    end
    local retained = math.max(0, maximum - M.damerau_levenshtein(left, right))
    return math.floor(retained * M.PERFECT_SCORE / maximum)
end

function M.candidate_score(query, candidate)
    local edit_score = M.similarity_score(query, candidate)
    if utf8.len(query) < M.PREFIX_MIN_CODE_POINTS then
        return edit_score
    end
    if candidate:sub(1, #query + 1) == query .. " " then
        return math.max(edit_score, M.TOKEN_PREFIX_SCORE)
    end
    if candidate:sub(1, #query) == query then
        return math.max(edit_score, M.PARTIAL_PREFIX_SCORE)
    end
    return edit_score
end

function M.glob_matches(pattern, candidate)
    local pattern_points = code_points(pattern)
    local candidate_points = code_points(candidate)
    local previous = { true }
    for index = 2, #candidate_points + 1 do
        previous[index] = false
    end

    for _, pattern_point in ipairs(pattern_points) do
        local current = {}
        current[1] = pattern_point == 0x2a and previous[1] or false
        for candidate_index = 1, #candidate_points do
            local column = candidate_index + 1
            if pattern_point == 0x2a then
                current[column] = previous[column] or current[column - 1]
            elseif pattern_point == 0x3f then
                current[column] = previous[column - 1]
            else
                current[column] = previous[column - 1]
                    and pattern_point == candidate_points[candidate_index]
            end
        end
        previous = current
    end
    return previous[#candidate_points + 1]
end

local states = setmetatable({}, { __mode = "k" })
local Dictionary = {}
Dictionary.__index = Dictionary
Dictionary.__newindex = function()
    error("attempt to modify an immutable dictionary", 2)
end
Dictionary.__metatable = "immutable"

function Dictionary.new(source_entries)
    local entries = {}
    local index = {}
    for position = 1, #source_entries do
        local source = source_entries[position]
        local entry = model.entry(
            source.word,
            source.part_of_speech,
            source.definition,
            source.references
        )
        entries[position] = entry
        index[position] = {
            entry = entry,
            normalized = normalizer.normalize(entry.word),
        }
    end
    table.sort(index, function(left, right)
        return M.compare_code_points(left.normalized, right.normalized) < 0
    end)
    for position = 2, #index do
        if index[position - 1].normalized == index[position].normalized then
            error("dictionary contains duplicate normalized headwords", 2)
        end
    end

    local self = setmetatable({}, Dictionary)
    states[self] = {
        entries = model.array(entries),
        index = index,
    }
    return self
end

function Dictionary:entries()
    return states[self].entries
end

function Dictionary:random_with(choose_index)
    local entries = states[self].entries
    if #entries == 0 then
        return nil
    end
    local index = choose_index(#entries)
    if type(index) ~= "number" or index % 1 ~= 0 or index < 0 or index >= #entries then
        return nil
    end
    return entries[index + 1]
end

function Dictionary:random()
    return self:random_with(function(bound)
        return math.random(0, bound - 1)
    end)
end

function Dictionary:resolve_with(request, choose_index)
    if request.type == "search" then
        return self:search(request.query)
    end
    local entry = self:random_with(choose_index)
    if entry == nil then
        return model.not_found
    end
    return model.found(entry, "random", nil)
end

function Dictionary:resolve(request)
    return self:resolve_with(request, function(bound)
        return math.random(0, bound - 1)
    end)
end

local function lower_bound(index, query)
    local lower = 1
    local upper = #index + 1
    while lower < upper do
        local middle = lower + math.floor((upper - lower) / 2)
        if M.compare_code_points(index[middle].normalized, query) < 0 then
            lower = middle + 1
        else
            upper = middle
        end
    end
    return lower
end

function Dictionary:search(query)
    if query:find("*", 1, true) or query:find("?", 1, true) then
        return self:search_glob(query)
    end

    local normalized_query = normalizer.normalize(query)
    if normalized_query == "" then
        return model.not_found
    end
    local state = states[self]
    local exact_position = lower_bound(state.index, normalized_query)
    if exact_position <= #state.index
        and state.index[exact_position].normalized == normalized_query
    then
        return model.found(state.index[exact_position].entry, "exact", M.PERFECT_SCORE)
    end

    local ranked = {}
    for position, indexed in ipairs(state.index) do
        ranked[position] = {
            entry = indexed.entry,
            score = M.candidate_score(normalized_query, indexed.normalized),
        }
    end
    table.sort(ranked, function(left, right)
        if left.score ~= right.score then
            return left.score > right.score
        end
        return M.compare_code_points(left.entry.word, right.entry.word) < 0
    end)

    local qualified = 0
    while qualified < #ranked and ranked[qualified + 1].score >= M.QUALIFYING_SCORE do
        qualified = qualified + 1
    end
    if qualified == 1 then
        local candidate = ranked[1]
        return model.found(candidate.entry, "high_confidence", candidate.score)
    end
    if qualified == 0 then
        return model.not_found
    end

    local suggestions = {}
    for position = 1, qualified do
        local candidate = ranked[position]
        suggestions[#suggestions + 1] = model.suggestion(
            candidate.entry,
            "medium",
            candidate.score
        )
    end
    local last = math.min(#ranked, qualified + M.LOW_SUGGESTION_COUNT)
    for position = qualified + 1, last do
        local candidate = ranked[position]
        suggestions[#suggestions + 1] = model.suggestion(
            candidate.entry,
            "low",
            candidate.score
        )
    end
    return model.did_you_mean(suggestions)
end

function Dictionary:search_glob(query)
    local pattern = normalizer.normalize_glob(query)
    if pattern == "" then
        return model.not_found
    end
    local matches = {}
    for _, indexed in ipairs(states[self].index) do
        if M.glob_matches(pattern, indexed.normalized) then
            matches[#matches + 1] = indexed.entry
        end
    end
    if #matches == 0 then
        return model.not_found
    end
    if #matches == 1 then
        return model.found(matches[1], "glob", M.PERFECT_SCORE)
    end
    local suggestions = {}
    for position, entry in ipairs(matches) do
        suggestions[position] = model.suggestion(entry, "medium", M.PERFECT_SCORE)
    end
    return model.did_you_mean(suggestions)
end

M.Dictionary = Dictionary

return M

File: impl/typescript/src/main/ts/dictionary.ts
import {
    NOT_FOUND,
    didYouMean,
    found,
    makeEntry,
    suggestion,
    type Entry,
    type Outcome,
    type Request,
} from "./model";
import { normalize, normalizeGlob } from "./normalize";

export const PERFECT_SCORE = 1000;
export const QUALIFYING_SCORE = 700;
export const LOW_SUGGESTION_COUNT = 2;
export const TOKEN_PREFIX_SCORE = 900;
export const PARTIAL_PREFIX_SCORE = 750;
export const PREFIX_MIN_CODE_POINTS = 4;

interface IndexedEntry {
    readonly entry: Entry;
    readonly normalized: string;
}

interface ScoredCandidate {
    readonly entry: Entry;
    readonly score: number;
}

export class Dictionary {
    readonly #entries: readonly Entry[];
    readonly #index: readonly IndexedEntry[];

    constructor(entries: readonly Entry[]) {
        this.#entries = Object.freeze(entries.map((entry) => makeEntry(
            entry.word,
            entry.partOfSpeech,
            entry.definition,
            entry.references,
        )));
        const index = this.#entries.map((entry) => Object.freeze({
            entry,
            normalized: normalize(entry.word),
        }));
        index.sort((left, right) => compareCodePoints(left.normalized, right.normalized));
        for (let position = 1; position < index.length; position++) {
            if (index[position - 1]!.normalized === index[position]!.normalized) {
                throw new Error("dictionary contains duplicate normalized headwords");
            }
        }
        this.#index = Object.freeze(index);
    }

    entries(): readonly Entry[] {
        return this.#entries;
    }

    random(): Entry | undefined {
        return this.randomWith((bound) => Math.floor(Math.random() * bound));
    }

    randomWith(chooseIndex: (exclusiveUpperBound: number) => number): Entry | undefined {
        if (this.#entries.length === 0) {
            return undefined;
        }
        const index = chooseIndex(this.#entries.length);
        if (!Number.isInteger(index) || index < 0 || index >= this.#entries.length) {
            return undefined;
        }
        return this.#entries[index];
    }

    resolve(request: Request): Outcome {
        return this.resolveWith(request, (bound) => Math.floor(Math.random() * bound));
    }

    resolveWith(
        request: Request,
        chooseIndex: (exclusiveUpperBound: number) => number,
    ): Outcome {
        if (request.type === "search") {
            return this.search(request.query);
        }
        const entry = this.randomWith(chooseIndex);
        return entry === undefined ? NOT_FOUND : found(entry, "random", null);
    }

    search(query: string): Outcome {
        if (query.includes("*") || query.includes("?")) {
            return this.searchGlob(query);
        }

        const normalizedQuery = normalize(query);
        if (normalizedQuery.length === 0) {
            return NOT_FOUND;
        }
        const exactPosition = this.lowerBound(normalizedQuery);
        if (exactPosition < this.#index.length
                && this.#index[exactPosition]!.normalized === normalizedQuery) {
            return found(this.#index[exactPosition]!.entry, "exact", PERFECT_SCORE);
        }

        const ranked: ScoredCandidate[] = this.#index.map((indexed) => ({
            entry: indexed.entry,
            score: candidateScore(normalizedQuery, indexed.normalized),
        }));
        ranked.sort((left, right) => right.score - left.score
            || compareCodePoints(left.entry.word, right.entry.word));

        let qualified = 0;
        while (qualified < ranked.length && ranked[qualified]!.score >= QUALIFYING_SCORE) {
            qualified++;
        }
        if (qualified === 1) {
            const candidate = ranked[0]!;
            return found(candidate.entry, "high_confidence", candidate.score);
        }
        if (qualified === 0) {
            return NOT_FOUND;
        }

        const suggestions = ranked.slice(0, qualified).map((candidate) =>
            suggestion(candidate.entry, "medium", candidate.score));
        for (const candidate of ranked.slice(qualified, qualified + LOW_SUGGESTION_COUNT)) {
            suggestions.push(suggestion(candidate.entry, "low", candidate.score));
        }
        return didYouMean(suggestions);
    }

    private searchGlob(query: string): Outcome {
        const pattern = normalizeGlob(query);
        if (pattern.length === 0) {
            return NOT_FOUND;
        }
        const matches = this.#index.filter((indexed) =>
            globMatches(pattern, indexed.normalized));
        if (matches.length === 0) {
            return NOT_FOUND;
        }
        if (matches.length === 1) {
            return found(matches[0]!.entry, "glob", PERFECT_SCORE);
        }
        return didYouMean(matches.map((indexed) =>
            suggestion(indexed.entry, "medium", PERFECT_SCORE)));
    }

    private lowerBound(query: string): number {
        let lower = 0;
        let upper = this.#index.length;
        while (lower < upper) {
            const middle = lower + Math.floor((upper - lower) / 2);
            if (compareCodePoints(this.#index[middle]!.normalized, query) < 0) {
                lower = middle + 1;
            } else {
                upper = middle;
            }
        }
        return lower;
    }
}

export function similarityScore(left: string, right: string): number {
    const maximum = Math.max(Array.from(left).length, Array.from(right).length);
    if (maximum === 0) {
        return PERFECT_SCORE;
    }
    const retained = Math.max(0, maximum - damerauLevenshtein(left, right));
    return Math.floor(retained * PERFECT_SCORE / maximum);
}

export function candidateScore(query: string, candidate: string): number {
    const editScore = similarityScore(query, candidate);
    if (Array.from(query).length < PREFIX_MIN_CODE_POINTS) {
        return editScore;
    }
    if (candidate.startsWith(`${query} `)) {
        return Math.max(editScore, TOKEN_PREFIX_SCORE);
    }
    if (candidate.startsWith(query)) {
        return Math.max(editScore, PARTIAL_PREFIX_SCORE);
    }
    return editScore;
}

export function globMatches(pattern: string, candidate: string): boolean {
    const patternPoints = Array.from(pattern);
    const candidatePoints = Array.from(candidate);
    let previous = new Array<boolean>(candidatePoints.length + 1).fill(false);
    previous[0] = true;
    for (const patternPoint of patternPoints) {
        const current = new Array<boolean>(candidatePoints.length + 1).fill(false);
        if (patternPoint === "*") {
            current[0] = previous[0]!;
        }
        for (let column = 1; column <= candidatePoints.length; column++) {
            if (patternPoint === "*") {
                current[column] = previous[column]! || current[column - 1]!;
            } else if (patternPoint === "?") {
                current[column] = previous[column - 1]!;
            } else {
                current[column] = previous[column - 1]!
                    && patternPoint === candidatePoints[column - 1];
            }
        }
        previous = current;
    }
    return previous[candidatePoints.length]!;
}

export function damerauLevenshtein(left: string, right: string): number {
    const leftPoints = Array.from(left);
    const rightPoints = Array.from(right);
    let previousPrevious = new Array<number>(rightPoints.length + 1).fill(0);
    let previous = Array.from({ length: rightPoints.length + 1 }, (_, index) => index);
    for (let leftIndex = 0; leftIndex < leftPoints.length; leftIndex++) {
        const current = new Array<number>(rightPoints.length + 1).fill(0);
        current[0] = leftIndex + 1;
        for (let rightIndex = 0; rightIndex < rightPoints.length; rightIndex++) {
            const column = rightIndex + 1;
            const substitution = leftPoints[leftIndex] === rightPoints[rightIndex] ? 0 : 1;
            current[column] = Math.min(
                current[column - 1]! + 1,
                previous[column]! + 1,
                previous[column - 1]! + substitution,
            );
            if (leftIndex > 0 && rightIndex > 0
                    && leftPoints[leftIndex] === rightPoints[rightIndex - 1]
                    && leftPoints[leftIndex - 1] === rightPoints[rightIndex]) {
                current[column] = Math.min(
                    current[column]!,
                    previousPrevious[column - 2]! + 1,
                );
            }
        }
        previousPrevious = previous;
        previous = current;
    }
    return previous[rightPoints.length]!;
}

export function compareCodePoints(left: string, right: string): number {
    const leftIterator = left[Symbol.iterator]();
    const rightIterator = right[Symbol.iterator]();
    while (true) {
        const leftNext = leftIterator.next();
        const rightNext = rightIterator.next();
        if (leftNext.done || rightNext.done) {
            if (leftNext.done && rightNext.done) {
                return 0;
            }
            return leftNext.done ? -1 : 1;
        }
        const difference = leftNext.value.codePointAt(0)! - rightNext.value.codePointAt(0)!;
        if (difference !== 0) {
            return difference;
        }
    }
}
